File: Backend/src/utils/connectionManager.js
const EventEmitter = require('events');

const DEFAULT_TIMEOUT_MS = 60000;

class ConnectionManager extends EventEmitter {
    constructor() {
        super();
        this.states = new Map();
    }

    _initState(serviceName) {
        if (!this.states.has(serviceName)) {
            this.states.set(serviceName, { status: 'disconnected' });
        }
    }

    setStatus(serviceName, status) {
        this._initState(serviceName);
        const state = this.states.get(serviceName);
        
        if (state.status !== status) {
            state.status = status;
            console.log(`[ConnectionManager] ${serviceName} is now ${status}`);
            if (status === 'connected') {
                this.emit(`${serviceName}:connected`);
            }
        }
    }

    getStatus(serviceName) {
        this._initState(serviceName);
        return this.states.get(serviceName).status;
    }

    isConnected(serviceName) {
        return this.getStatus(serviceName) === 'connected';
    }

    async ensureConnection(serviceName, triggerConnectFn = null, timeoutMs = DEFAULT_TIMEOUT_MS) {
        this._initState(serviceName);
        const state = this.states.get(serviceName);

        if (state.status === 'connected') {
            return; // Fast path
        }

        console.log(`[ConnectionManager] Operation paused for ${serviceName}. Current status: ${state.status}`);

        if (state.status === 'disconnected') {
            this.setStatus(serviceName, 'connecting');
            if (triggerConnectFn) {
                try {
                    await triggerConnectFn();
                } catch (error) {
                    console.log(`[ConnectionManager] Connection trigger failed for ${serviceName}:`, error.message);
                }
            }
        }

        // Wait until it becomes connected, with timeout to prevent infinite blocking
        if (this.states.get(serviceName).status !== 'connected') {
            await new Promise((resolve, reject) => {
                const onConnect = () => {
                    clearTimeout(timer);
                    this.removeListener(`${serviceName}:connected`, onConnect);
                    resolve();
                };

                const timer = setTimeout(() => {
                    this.removeListener(`${serviceName}:connected`, onConnect);
                    reject(new Error(`[ConnectionManager] Timeout: ${serviceName} did not reconnect within ${timeoutMs}ms`));
                }, timeoutMs);

                this.on(`${serviceName}:connected`, onConnect);
            });
            console.log(`[ConnectionManager] Operation resumed for ${serviceName}.`);
        }
    }
}

module.exports = new ConnectionManager();

File: AI/text/utils/connection_manager.py
import time
import threading
import logging

logger = logging.getLogger(__name__)

class ConnectionManager:
    def __init__(self):
        self.states = {}
        self.lock = threading.RLock()
        
    def _init_state(self, service_name):
        with self.lock:
            if service_name not in self.states:
                self.states[service_name] = 'disconnected'
                
    def set_status(self, service_name, status):
        with self.lock:
            self._init_state(service_name)
            if self.states[service_name] != status:
                self.states[service_name] = status
                logger.info(f"[ConnectionManager] {service_name} is now {status}")
                    
    def get_status(self, service_name):
        with self.lock:
            self._init_state(service_name)
            return self.states[service_name]
            
    def ensure_connection(self, service_name, ping_fn=None):
        self._init_state(service_name)
        
        if self.get_status(service_name) == 'connected':
            return
            
        logger.info(f"[ConnectionManager] Operation paused for {service_name}. Current status: {self.get_status(service_name)}")
        
        with self.lock:
            if self.states[service_name] == 'disconnected':
                self.states[service_name] = 'connecting'
                
        while True:
            status = self.get_status(service_name)
            if status == 'connected':
                break
                
            if status == 'connecting':
                if ping_fn:
                    try:
                        ping_fn()
                        self.set_status(service_name, 'connected')
                        logger.info(f"[ConnectionManager] Operation resumed for {service_name}.")
                        break
                    except Exception as e:
                        time.sleep(2)
                else:
                    time.sleep(1)

connection_manager = ConnectionManager()
